tremor-script/src/event_payload.rs

```diff
@@ -0,0 +1,285 @@
+///   - Each Vec is pinned to ensure the underlying data isn't moved
+///   - Each Pin is in a Arc so we can clone the data without with both clones
+///     still pointing to the underlying pin.
+///
+/// It is essential to never access the parts of the struct outside of it's
+/// implementation! This will void all warenties and likely lead to errors.
+///
+/// They **must** remain private. All interactions with them have to be guarded
+/// by the implementation logic to ensure they remain sane.
+///
+#[derive(Clone, Default)]
+pub struct EventPayload {
+    /// The vector of raw input values
+    raw: Vec<Arc<Pin<Vec<u8>>>>,
+    structured: ValueAndMeta<'static>,
+}
+
+impl std::fmt::Debug for EventPayload {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.structured.fmt(f)
+    }
+}
+
+impl EventPayload {
+    /// Borrows the structured part of the struct.
+    #[must_use]
+    pub fn suffix(&self) -> &ValueAndMeta {
+        &self.structured
+    }
+
+    /// a function to turn it into a value and metadata set.
+    ///
+    /// The return can reference the the data it gets passed
+    /// in the function.
+    ///
+    /// Internally the lifetime will be bound to the raw part
+    /// of the struct.
+    #[must_use]
+    pub fn new<F>(raw: Vec<u8>, f: F) -> Self
+    where
+        F: for<'head> FnOnce(&'head mut [u8]) -> ValueAndMeta<'head>,
+    {
+        let mut raw = Pin::new(raw);
+        let structured = f(raw.as_mut().get_mut());
+        // This is where the magic happens
+        let structured: ValueAndMeta<'static> = unsafe { mem::transmute(structured) };
+        let raw = vec![Arc::new(raw)];
+        Self { raw, structured }
+    }
+
+    /// Creates a new Payload with a given byte vector and
+    /// a function to turn it into a value and metadata set.
+    ///
+    /// The return can reference the the data it gets passed
+    /// in the function.
+    ///
+    /// Internally the lifetime will be bound to the raw part
+    /// of the struct.
+    ///
+    /// # Errors
+    /// errors if the conversion function fails
+    pub fn try_new<E, F>(raw: Vec<u8>, f: F) -> std::result::Result<Self, E>
+    where
+        F: for<'head> FnOnce(&'head mut [u8]) -> std::result::Result<ValueAndMeta<'head>, E>,
+    {
+        let mut raw = Pin::new(raw);
+        let structured = f(raw.as_mut().get_mut())?;
+        // This is where the magic happens
+        let structured: ValueAndMeta<'static> = unsafe { mem::transmute(structured) };
+        let raw = vec![Arc::new(raw)];
+        Ok(Self { raw, structured })
+    }
+
+    /// Named after the original rental struct for easy rewriting.
+    ///
+    /// Borrows the borrowed (liftimed) part of the self referential struct
+    /// and calls the provided function with a reference to it
+    pub fn rent<F, R>(&self, f: F) -> R
+    where
+        F: for<'iref, 'head> FnOnce(&'iref ValueAndMeta<'head>) -> R,
+        R: ,
```

Comment:
what does this mean, can you leave this out?

Reply:
I genuenly am not sure, I copied the signature from the rental, and I'm unsure if it means nothing or means `R can't include 'head or 'iref` so I aired on the side of caution